diag-client-lib/appl/src/dcm/conversation/dm_conversation.cpp

```diff
@@ -253,30 +253,46 @@ DmConversation::IndicateMessage(uds_transport::UdsMessage::Address source_addr,
     // Check for size, else kIndicationOverflow
     if (size <= rx_buffer_size_) {
       // Check for pending response
-      // payload = 0x7F XX 0x78
-      if (payload_info[2U] == 0x78) {
-        logger::DiagClientLogger::GetDiagClientLogger().GetLogger().LogInfo(
-            __FILE__, __LINE__, "", [&](std::stringstream &msg) {
-              msg << "'" << conversation_name_ << "'"
-                  << "-> "
-                  << "Diagnostic pending response received in Conversation";
-            });
-        ret_val.first = uds_transport::UdsTransportProtocolMgr::IndicationResult::kIndicationPending;
-        conversation_state_.GetConversationStateContext().TransitionTo(ConversationState::kDiagRecvdPendingRes);
+      if (size >= 3U) {
+        // payload = 0x7F XX 0x78
+        if (payload_info[2U] == 0x78) {
+          logger::DiagClientLogger::GetDiagClientLogger().GetLogger().LogInfo(
+              __FILE__, __LINE__, "", [&](std::stringstream &msg) {
+                msg << "'" << conversation_name_ << "'"
+                    << "-> "
+                    << "Diagnostic pending response received in Conversation";
+              });
+          ret_val.first = uds_transport::UdsTransportProtocolMgr::IndicationResult::kIndicationPending;
+          conversation_state_.GetConversationStateContext().TransitionTo(ConversationState::kDiagRecvdPendingRes);
+        } else {
+          logger::DiagClientLogger::GetDiagClientLogger().GetLogger().LogDebug(
+              __FILE__, __LINE__, "", [this](std::stringstream &msg) {
+                msg << "'" << conversation_name_ << "'"
+                    << "-> "
+                    << "Diagnostic final response received in Conversation";
+              });
+          // positive or negative response, provide valid buffer
+          // resize the global rx buffer
+          payload_rx_buffer.resize(size);
+          ret_val.first = uds_transport::UdsTransportProtocolMgr::IndicationResult::kIndicationOk;
+          ret_val.second = std::move(std::make_unique<diag::client::uds_message::DmUdsMessage>(
+              source_address_, target_address_, "", payload_rx_buffer));
+          conversation_state_.GetConversationStateContext().TransitionTo(ConversationState::kDiagRecvdFinalRes);
+        }
```

Comment:
```suggestion
        // payload = 0x7F XX 0x78
        if (payload_info[0U] == 0x7F && payload_info[2U] == 0x78) {
          logger::DiagClientLogger::GetDiagClientLogger().GetLogger().LogInfo(
              __FILE__, __LINE__, "", [&](std::stringstream &msg) {
                msg << "'" << conversation_name_ << "'"
                    << "-> "
                    << "Diagnostic pending response received in Conversation";
              });
          ret_val.first = uds_transport::UdsTransportProtocolMgr::IndicationResult::kIndicationPending;
          conversation_state_.GetConversationStateContext().TransitionTo(ConversationState::kDiagRecvdPendingRes);
        } else {
          logger::DiagClientLogger::GetDiagClientLogger().GetLogger().LogDebug(
              __FILE__, __LINE__, "", [this](std::stringstream &msg) {
                msg << "'" << conversation_name_ << "'"
                    << "-> "
                    << "Diagnostic final response received in Conversation";
              });
          // positive or negative response, provide valid buffer
          // resize the global rx buffer
          payload_rx_buffer.resize(size);
          ret_val.first = uds_transport::UdsTransportProtocolMgr::IndicationResult::kIndicationOk;
          ret_val.second = std::move(std::make_unique<diag::client::uds_message::DmUdsMessage>(
              source_address_, target_address_, "", payload_rx_buffer));
          conversation_state_.GetConversationStateContext().TransitionTo(ConversationState::kDiagRecvdFinalRes);
        }
```

Reply:
hello Iandiehard, I still worry about the size of payload_info. if the size of payload_info is less than 2 bytes, this will error again

Reply:
As per my knowledge, UDS standard says Negative Response should always have 3 bytes (7F XX YY) where XX is Service Id and YY is NRC code. So its unlikely that we receive 7F with 2 bytes of data as its against the UDS standard.

If payload is minimum two bytes then it has to be a positive response.